### PR TITLE
Adding the read/write functions

### PR DESCRIPTION
**Pull request by AFulleringer**
Added the read and write functions - need to be validated with actual prompts.

### CODE REVIEW

**Review by gebelangsn — APPROVED** [review]
LGTM! maybe wait for Kyle to merge his branch, then pull from master and push to this branch before merging... (just to make sure no conflicts)

File: src/affordance/tasks/tools.py [hunk 1]
```diff
@@ -25,7 +25,49 @@
 - (str) the output
 - (str) runtime details to be logged
 """
+import pickle
```
Comment (kylrth) [inline src/affordance/tasks/tools.py]: Put this import up with the other imports so it can be sorted in

File: src/affordance/tasks/tools.py [hunk 2]
```diff
@@ -25,7 +25,49 @@
...
 
+def get_run_dict(run_title):
+    try:
+        with open(f'{run_title}.pickle', 'rb') as handle:
+            run_dict = pickle.load(handle)
+    except FileNotFoundError:
+        run_dict = {}
+        print('run dict not found, creating a new one!')
```
Comment (kylrth) [inline src/affordance/tasks/tools.py]:
```suggestion
        print("creating a new dict for run:", run_title)
```

File: src/affordance/tasks/tools.py [hunk 3]
```diff
@@ -25,7 +25,49 @@
...
+    return run_dict
+
+def save_run_dict(run_dict, run_title):
+    with open(f'{run_title}.pickle', 'wb') as handle:
```
Comment (kylrth) [inline src/affordance/tasks/tools.py]: We should make it possible to choose the directory for this with a flag. Or to save time we should just change it to somewhere not in cwd, maybe `./kvstore` or something?

File: src/affordance/tasks/tools.py [hunk 4]
```diff
@@ -25,7 +25,49 @@
...
+        pickle.dump(run_dict, handle, protocol=pickle.HIGHEST_PROTOCOL)
+    return
+
+
+def read(key, run_title) -> tuple(str, str):
```
Comment (kylrth) [inline src/affordance/tasks/tools.py]: The way the code works currently, the first argument is the input text for the command, and the second argument is the output from the previous command. We should probably change the way the tools are called across the board, so they all have three arguments: a `run_title`, a `cmd_input`, and a `prev_output`. Mind adding that change into this PR?

File: src/affordance/tasks/tools.py [hunk 5]
```diff
@@ -25,7 +25,49 @@
...
+    read_write_dict = get_run_dict(run_title)
+    if key in read_write_dict:
+        return read_write_dict[key], None
+    #TODO finalize key not found behaviour
+    else:
+        return ', '.join(read_write_dict.keys()), None
+    return f"<key \'{key}\' did not exist>", None  # TODO: return a useful log message
```
Comment (kylrth) [inline src/affordance/tasks/tools.py]: These two returns are redundant, and whichever one you keep doesn't need to be in an else block

File: src/affordance/tasks/tools.py [hunk 6]
```diff
@@ -25,7 +25,49 @@
...
+
+def parse_instruction(instruction):
+    # parse `instruction` = (first day creation) God created the earth, light, and dark.
+    split = instruction.split(")", 1)
+    key = split[0][1:]  # remove "("
+    text = split[1].strip()
+    return key, text
+
+def write(instruction, run_title) -> tuple(str, str):
+
```
Comment (kylrth) [inline src/affordance/tasks/tools.py:72]: run the `black` formatter real quick

File: src/affordance/tasks/tools.py [hunk 7]
```diff
@@ -25,7 +25,49 @@
...
+    key, text = parse_instruction(instruction)
+    # key = first day creation
+    # text = God created the earth, light, and dark.
```
Comment (kylrth) [inline src/affordance/tasks/tools.py]: Maybe let's move some of these example comments into docstrings or something

**Review by kylrth — CHANGES_REQUESTED** [review]
I've got some suggestions! Isn't code review fun

**Comment by kylrth** [comment]
Oh also don't forget to add this bit:

```python
    "[read]": read,
    "[write]": write,
    "[list keys]": list_keys,
```

In the tool dict

**Comment by AFulleringer** [comment]
Somehow the changes to add run_title to visit and its lineage ended up in 'reduce cost of debug config' - just a heads up on where to look for that.

File: .vscode/launch.json [hunk 8]
```diff
@@ -15,6 +15,27 @@
                 "nl_program",
                 "--num_train_examples",
                 "1",
+<<<<<<< HEAD
```
Comment (kylrth) [inline .vscode/launch.json]: Haha somehow a merge conflict ended up here. It sounds like you have `commit`-ment issues!

I'm going to add one more commit to remove some of this stuff, and then when we're finished we can squash everything down into a single commit for posterity.

File: src/affordance/tasks/arxiv_summarization.py [hunk 9]
```diff
@@ -0,0 +1,384 @@
+import argparse
```
Comment (kylrth) [inline src/affordance/tasks/arxiv_summarization.py]: I actually think it's better to have a PR contain just one new thing at a time. Better to leave Georges' stuff in his branch

File: src/affordance/tasks/sequential_interpreter.py [hunk 10]
```diff
@@ -43,10 +43,12 @@ def program_ends(program: str) -> bool:
         """Return true if the program contains [EOQ] and a final answer."""
         return "[EOQ]\nAns:" in program
 
-    def batch_visit(self, prefixes: List[str], programs: List[str]) -> List[str]:
+    def batch_visit(
```
Comment (kylrth) [inline src/affordance/tasks/sequential_interpreter.py:47]: Sometimes this function is called with a batch across multiple separate evaluations, so we probably need to make `run_title` be a `List[str]`.

File: src/affordance/tasks/sequential_interpreter.py [hunk 11]
```diff
@@ -57,7 +59,9 @@ def shorten_prefix(self, prefix: str, to_remove: int) -> str:
         # split along demonstrations
         individual_programs = prefix.split("\n----\n")
 
-        shifted_programs = individual_programs[: -(to_remove + 1)] + [individual_programs[-1]]
+        shifted_programs = individual_programs[: -(to_remove + 1)] + [
```
Comment (kylrth) [inline src/affordance/tasks/sequential_interpreter.py]: looks like I forgot to mention that I've been using `black --line-length 100`, which allows these lines to be longer. I'll fix that

File: src/affordance/tasks/sequential_interpreter.py [hunk 12]
```diff
@@ -187,7 +199,9 @@ def visit(self, prefix: str, program: str) -> str:
                     break
 
                 stack_trace.append(
-                    StacktraceItem(command, tool_func, tool_output, tool_details, program, p)
+                    StacktraceItem(
```
Comment (kylrth) [inline src/affordance/tasks/sequential_interpreter.py:193]: We should log the run_title too

**Comment by kylrth** [comment]
Alright once you make those few last changes, the last thing we need is to squash:

```sh
git fetch
git reset --soft origin/master
git commit -m "<a nice summary>"
git push -f
```

**Review by kylrth — APPROVED** [review]

File: .github/workflows/black.yml [hunk 13]
```diff
@@ -2,9 +2,9 @@ name: code formatting
 
 on:
   push:
-    branches: [ master ]
+    branches: [ main ]
   pull_request:
-    branches: [ master ]
+    branches: [ main ]
```
Comment (kylrth) [inline .github/workflows/black.yml]: You'll want to keep this

**Review by kylrth — CHANGES_REQUESTED** [review]

**Review by kylrth — APPROVED** [review]

**Pull request merged**